clis/jenkins-shipper/main.go

```diff
@@ -0,0 +1,509 @@
+			}
+		}
+	}
+
+	return stepsErrors, nil
+}
+
+func fetchAndDefaultTestsErrors(url string) (*gabs.Container, error) {
+	json, err := fetchAndDefault(url, true)
+	if err != nil {
+		return nil, err
+	}
+
+	if json.Exists("code") && json.Path("code").Data().(int) == 404 {
+		return gabs.New().Array()
+	}
+
+	for _, child := range json.Children() {
+		normaliseTests(child)
+	}
+
+	return json, nil
+}
+
+func fetchAndPrepareArtifactsInfo(url string) (*gabs.Container, error) {
+	json, err := fetchAndDefault(url, true)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, child := range json.Children() {
+		normaliseArtifacts(child)
+	}
+
+	return json, nil
+}
+
+func fetchAndPrepareBuildInfo(url string) (*gabs.Container, error) {
+	json, err := fetchAndDefault(url, false)
+	if err != nil {
+		return nil, err
+	}
+
+	normaliseBuild(jenkinsURL, json)
+
+	return json, nil
+}
+
+func fetchAndPrepareBuildReport(url string, isList bool) (*gabs.Container, error) {
+	json, err := fetchAndDefault(url, isList)
+	if err != nil {
+		return nil, err
+	}
+
+	normaliseArtifacts(json)
+	normaliseBuildReport(json)
+
+	for _, child := range json.Children() {
+		normaliseChangeset(child)
+	}
+
+	return json, nil
+}
+
+func fetchAndPrepareTestCoverageReport(url string) (*gabs.Container, error) {
+	json, err := fetch(url)
+	if err != nil {
+		// in the case there is not code coverage report, we return an empty object
+		if strings.EqualFold(err.Error(), "GET request failed with 404") {
+			fmt.Printf(">> code coverage not found at %s: %v", url, err)
+			return gabs.New(), nil
+		}
+
+		return nil, err
+	}
+
+	coverage := normaliseCoberturaSummary(json)
+	return coverage, nil
+}
+
+func fetchAndPrepareTestsInfo(url string) (*gabs.Container, error) {
+	json, err := fetchAndDefault(url, true)
+	if err != nil {
+		return nil, err
+	}
+
+	/*
+		Tests json response differs when there were tests executed in
+		the pipeline, otherwise it returns:
+		{ message: "no tests", code: 404, errors: [] }
+	*/
+	if json.Exists("code") && json.Path("code").Data().(int) == 404 {
+		return gabs.New(), nil
+	}
+
+	for _, child := range json.Children() {
+		normaliseTestsWithoutStacktrace(child)
+	}
+
+	return json, nil
+}
+
+func fetchAndPrepareTestSummaryReport(url string, testJSON *gabs.Container) (*gabs.Container, error) {
+	json, err := fetchAndDefault(url, true)
+	if err != nil {
+		// BlueOcean might return 500 in some scenarios. If so, let's parse the tests entrypoint
+		if testJSON != nil {
+			total := len(testJSON.Children())
+			passed := 0
+			failed := 0
+			skipped := 0
+			for _, test := range testJSON.Children() {
+				status := test.Path("status").Data().(string)
+				switch status {
+				case "FAILED":
+					failed++
+				case "PASSED":
+					passed++
+				case "SKIPPED":
+					skipped++
+				}
+			}
+
+			json = gabs.New()
+			json.Set(total, "total")
+			json.Set(passed, "passed")
+			json.Set(failed, "failed")
+			json.Set(skipped, "skipped")
+		}
+
+		return nil, err
```

Comment:
IIUC, this manipulation not be returned here but later on?

```suggestion
```

Reply:
Mmm, you are right, but it's not a matter of removing the return there, but guarding it with the `else`